File: advent-of-code-2021/day2/Solution.java
import java.io.*;
import java.util.*;

public class Solution {
    public static void main(String[] args) throws Exception {
        File file = new File("E:\\Coding\\advent-of-code-2021\\day2\\input.txt");
        BufferedReader br = new BufferedReader(new FileReader(file));
        int position = 0;
        int depth = 0;
        int aim = 0;

        String s;
        while ((s = br.readLine()) != null) {
            char lastChar = s.charAt(s.length() - 1);
            char firstChar = s.charAt(0);
            int value = Character.getNumericValue(lastChar);

            if (firstChar == 'f') { // forward
                position += value;
                depth += (aim * value);
            } else if (firstChar == 'd') { // down
                aim += value;
            } else if (firstChar == 'u') { // up
                aim -= value;
            } else {
                System.out.println("Unexpected first character encountered");
            }
        }

        System.out.println(position * depth);
        br.close();
    }
}

File: advent-of-code-2021/day2/solution.py
def main():
    print("Part one: ", part_one())
    print("Part two: ", part_two())

def part_one():
    position = 0; depth = 0

    with open("input.txt") as file:
        while (line := file.readline().rstrip()):
            firstChar = line[0]
            lastChar = line[-1]
            value = int(lastChar)

            if (firstChar == 'f'):
                position += value
            elif (firstChar == 'd'):
                depth += value
            elif (firstChar == 'u'):
                depth -= value
            else:
                print("Unexpected first character encountered.")
    
    return position * depth

def part_two():
    position = 0; depth = 0; aim = 0

    with open("input.txt") as file:
        while (line := file.readline().rstrip()):
            firstChar = line[0]
            lastChar = line[-1]
            value = int(lastChar)

            if (firstChar == 'f'):
                position += value
                depth += (aim * value)
            elif (firstChar == 'd'):
                aim += value
            elif (firstChar == 'u'):
                aim -= value
            else:
                print("Unexpected first character encountered.")
    
    return position * depth


if __name__ == "__main__":
    main()
